### ISSUE TITLE
Add options to run hub and proxy pods on host network

### ISSUE THREAD

**Issue by mallman** [jupyterhub/zero-to-jupyterhub-k8s#2256]
### Proposed change
Add configuration options to the helm chart to deploy the hub and/or proxy pods on the host network.

### Alternative options
The alternative is the current configuration, which does not put the either the hub or the proxy on the host network. Rather, it creates a proxy-public service, which by default is of type "load balancer". It can also be configured as a "node port". However, the current configuration is (to the best of my ability) incompatible with the scenario where the notebook pods are deployed on the host network (see below).

### Who would use this feature?
This will greatly simplify and ease deployment when the notebook pods are configured to run on the host network. We need to run the notebook pods on the host network for compatibility with Spark. Spark requires a routable client address. When Spark runs outside the kubernetes cluster, the client cannot send an address to Spark that's routable back to the notebook pod.

### (Optional): Suggest a solution
Add the configuration options to the chart and set the value of the 'hostNetwork' pod spec as appropriate. There will likely be more work to do beyond that, like reconfiguring the public proxy service to route to the host address of the proxy instead of an internal address.

**Comment by manics** [comment]
Hi! Could you explain a bit more about why running the singleuser pods with host networking is incompatible with the standard services for the hub and proxy? It's possible the JupyterHub config would need to be modified to pass the external (add opposed to the internal k8s) address to the singleuser servers, but that should be a less drastic change than making everything use host networking.

**Comment by mallman** [comment]
My understanding is that the hub must send its URL to the notebook server. What configuration setting should I set the external URL in? This is what I'm considering:

c.JupyterHub.hub_connect_url
c.JupyterHub.hub_ip
c.JupyterHub.ip
c.JupyterHub.port

**Comment by mallman** [comment]
Let me expand on our situation. My understanding from https://jupyterhub.readthedocs.io/en/stable/reference/technical-overview.html#the-subsystems-hub-proxy-single-user-notebook-server is we need an external address for the browser. This is the proxy-public service. We have this set up with a NodePort because we are bare-metal and don't support the LoadBalancer configuration.

The proxy connects to the hub via its internal service connection. It also connects to the notebook server. Since we are running the notebook server pod on host networking, the proxy connects to its host address. The notebook server needs to connect to the Hub. Since it's on the host network, it needs a connection to an external address. So I suppose we set up the Hub service with a nodeport connection and configure the notebook servers to connect to the hub on that port.

If I understand correctly, I need to configure the hub and proxy services to use node ports in `values.yaml`. But I'm stumped on how to configure

c.JupyterHub.hub_connect_url
c.JupyterHub.hub_ip
c.JupyterHub.ip
c.JupyterHub.port
c.ConfigurableHTTPProxy.api_url

and maybe something else? I've tried various combinations without success. This is what my service configuration currently looks like

```
NAME           TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)          AGE
hub            NodePort    10.233.56.18    <none>        8081:30201/TCP   165m
proxy-api      ClusterIP   10.233.41.109   <none>        8001/TCP         165m
proxy-public   NodePort    10.233.14.221   <none>        80:30200/TCP     165m
```

Note I've also disabled the hub network policy because that was blocking packets from the external network.

**Comment by mallman** [comment]
I think the underlying issue is that the proxy and notebook server need a URL they can both use to connect to the hub. I think that's the `hub_connect_url` configuration setting. In this case, I suppose that would be based on the hub service node port of 30201. For example, we have a node with address `spark-poc-5360-105-027.prod.dc3`. Then `hub_connect_url` should be `http://spark-poc-5360-105-027.prod.dc3:30201/` and we should be able to curl that address, right? Yet we get intermittent connection timeouts that we've been unable to fathom:

```
[msa@dc3-5360-102-003 ~]$ curl -v http://spark-poc-5360-105-027.prod.dc3:30201/
* About to connect() to spark-poc-5360-105-027.prod.dc3 port 30201 (#0)
*   Trying 172.22.2.38...
* Connection timed out
* Failed connect to spark-poc-5360-105-027.prod.dc3:30201; Connection timed out
* Closing connection 0
curl: (7) Failed connect to spark-poc-5360-105-027.prod.dc3:30201; Connection timed out
```

The proxy also times out trying to connect to `http://spark-poc-5360-105-027.prod.dc3:30201/` and things go sour.

The only reliable mode of deployment we've found when deploying the notebook server pods on the host network is to deploy the hub and proxy pods with the host network, too.

**Comment by yuvipanda** [comment]
> Yet we get intermittent connection timeouts that we've been unable to fathom:

So if I understand this correctly, the current setup *does* work, but is unreliable, with intermittent failure. Putting the hub and proxy pod in hostNetwork makes this work *reliably*. is that correct?

My intuitive understanding of hostNetwork is that pods with hostNetwork should be able to talk to services pointing to pods without hostNetwork transparently, handled via `kube-proxy` (or similar). So I want to understand if this is either a (hub orr k8s) configuration issue, a reliability issue, or something fundamentally wrong about my understanding of hostNetwork.

**Comment by mallman** [comment]
Hi @yuvipanda. Let me reply in-line.

> So if I understand this correctly, the current setup does work, but is unreliable, with intermittent failure. Putting the hub and proxy pod in hostNetwork makes this work reliably. is that correct?

I'm reluctant to claim that the current setup works, because it's too unreliable to be usable. The connection timeouts occur more often than successful connections, so the setup really is unusable. And we've put days into trying to diagnose the underlying issue without success. I think our ops team is trying one more thing, but as of now we're still in the hole on this one. If we even knew what the problem is we could work on fixing it, but we haven't found that yet.

> My intuitive understanding of hostNetwork is that pods with hostNetwork should be able to talk to services pointing to pods without hostNetwork transparently, handled via kube-proxy (or similar).

Containers in pods with hostNetwork enabled are integrated into the host networking stack like a host-native process. They do not get access to or IP addresses on the internal Kubernetes network.

Does this answer your questions?

Cheers.

**Comment by mallman** [comment]
This has been addressed with PR #2306.
